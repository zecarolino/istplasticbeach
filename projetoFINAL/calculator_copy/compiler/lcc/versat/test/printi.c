void putchar(int ch) { asm("\trdw R11\n\taddi 2\n\twrw RB\n\trdwb\n\trdwb\n\twrw CPRT_BASE\n"); }
int _mul(int a, int b) {
	int c = 0;
	while (a) {
		if (a & 1) c += b;
		a >>= 1;
		b <<= 1;
	}
	return c;
}
int _div(int a, int b) {
	int c = 1, neg = 0, aux;
	if (!a) return 0;
	if (!b) return 2147483647; /* divide by zero */
	if (a < 0) a = -a, neg = 1;
	if (b < 0) b = -b, neg = 1 - neg;
	aux = b;
	if (b < a) {
		while ((aux << 1) - a < 0)
			aux <<= 1, c <<= 1;
		while (aux + b <= a)
			aux += b, c++;
	} else c = 0;
	if (neg) c = -c;
	return c;
}
int _mod(int a, int b) { return a - (a / b) * b; }

void printi(int n, int b) {
	int a;
	if (a = n / b) printi(a, b);
	putchar(n % b + '0');
}
int main() { printi(13579, 10); putchar('\n'); return 0; }
/* 13579 => 33756 cycles (4x times faster than basic _mod _div */
