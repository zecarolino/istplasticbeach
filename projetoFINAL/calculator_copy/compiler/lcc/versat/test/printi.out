VCD info: dumpfile xtop.vcd opened for output.
13579
0
Catched Trap at time  34109 clock cycles:
